src/Components/test/E2ETest/Tests/StatePersistenceTest.cs

```diff
@@ -0,0 +1,237 @@
+        {
+            // Navigate to a page without components first to make sure that we exercise rendering components
+            // with enhanced navigation on.
+            if (streaming == null)
+            {
+                Navigate($"subdir/persistent-state/page-no-components?render-mode={mode}&suppress-autostart");
+            }
+            else
+            {
+                Navigate($"subdir/persistent-state/page-no-components?render-mode={mode}&streaming-id={streaming}&suppress-autostart");
+            }
+            if (mode == "auto")
+            {
+                BlockWebAssemblyResourceLoad();
+            }
+            Browser.Click(By.Id("call-blazor-start"));
+            Browser.Click(By.Id("page-with-components-link"));
+        }
+        else
+        {
+            SuppressEnhancedNavigation(true);
+        }
+
+        if (mode != "auto")
+        {
+            RenderComponentsWithPersistentStateAndValidate(suppressEnhancedNavigation, mode, renderMode, streaming);
+        }
+        else
+        {
+            if (suppressEnhancedNavigation)
+            {
+                BlockWebAssemblyResourceLoad();
+            }
+            // For auto mode, validate that the state is persisted for both runtimes and is able
+            // to be loaded on server and wasm.
+            RenderComponentsWithPersistentStateAndValidate(suppressEnhancedNavigation, mode, renderMode, streaming, interactiveRuntime: "server");
+
+            UnblockWebAssemblyResourceLoad();
+            Browser.Navigate().Refresh();
+
+            RenderComponentsWithPersistentStateAndValidate(suppressEnhancedNavigation, mode, renderMode, streaming, interactiveRuntime: "wasm");
+        }
+    }
+
+    [Theory]
+    [InlineData((string)null)]
+    [InlineData("ServerStreaming")]
+    public async Task StateIsProvidedEveryTimeACircuitGetsCreated(string streaming)
+    {
+        var mode = "server";
+        if (streaming == null)
+        {
+            Navigate($"subdir/persistent-state/page-no-components?render-mode={mode}");
+        }
+        else
+        {
+            Navigate($"subdir/persistent-state/page-no-components?render-mode={mode}&streaming-id={streaming}");
+        }
+        Browser.Click(By.Id("page-with-components-link"));
+
+        RenderComponentsWithPersistentStateAndValidate(suppresEnhancedNavigation: false, mode, typeof(InteractiveServerRenderMode), streaming);
+        Browser.Click(By.Id("page-no-components-link"));
+        // Ensure that the circuit is gone.
+        await Task.Delay(1000);
+        Browser.Click(By.Id("page-with-components-link-and-state"));
+        RenderComponentsWithPersistentStateAndValidate(suppresEnhancedNavigation: false, mode, typeof(InteractiveServerRenderMode), streaming, stateValue: "other");
+    }
+
+    private void BlockWebAssemblyResourceLoad()
+    {
+        ((IJavaScriptExecutor)Browser).ExecuteScript("sessionStorage.setItem('block-load-boot-resource', 'true')");
```

Comment:
What is this used for? I can't see anything that reads the value.

Reply:
It forces the app to start on Blazor Server when auto is enabled on the page. Other similar tests follow this approach.